src/Sylius/Bundle/ApiBundle/ApiPlatform/Factory/MergingExtractorResourceMetadataFactory.php

```diff
@@ -0,0 +1,168 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\ApiPlatform\Factory;
+
+use ApiPlatform\Core\Exception\ResourceClassNotFoundException;
+use ApiPlatform\Core\Metadata\Extractor\ExtractorInterface;
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use ApiPlatform\Core\Metadata\Resource\ResourceMetadata;
+
+
+final class MergingExtractorResourceMetadataFactory implements ResourceMetadataFactoryInterface
+{
+    /** @var ExtractorInterface */
+    private $extractor;
+
+    /** @var ?ResourceMetadataFactoryInterface  */
+    private $decorated;
+
+    /** @var array */
+    private $defaults;
+
+    public function __construct(ExtractorInterface $extractor, ResourceMetadataFactoryInterface $decorated = null, array $defaults = [])
+    {
+        $this->extractor = $extractor;
+        $this->decorated = $decorated;
+        $this->defaults = $defaults + ['attributes' => []];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function create(string $resourceClass): ResourceMetadata
+    {
+        $parentResourceMetadata = null;
+        if ($this->decorated) {
+            try {
+                $parentResourceMetadata = $this->decorated->create($resourceClass);
+            } catch (ResourceClassNotFoundException $resourceNotFoundException) {
+                // Ignore not found exception from decorated factories
+            }
+        }
+
+        if (!(class_exists($resourceClass) || interface_exists($resourceClass)) || !$resource = $this->extractor->getResources()[$resourceClass] ?? false) {
+            return $this->handleNotFound($parentResourceMetadata, $resourceClass);
+        }
+
+        $resource['description'] = $resource['description'] ?? $this->defaults['description'] ?? null;
```

Comment:
Why the need for all these? Can't you just loop over an array of keys? Also I'd suggest to work on ResourceMetadata instead of on an array. 
Last but not least in API Platform 3.x `Core\Metadata\Resource` will be used instead of the ResourceMetadata and this will probably not be compatible. If you work on the ResourceMetadata there will be a bridge between ResourceMetadata and the new  `Core\Metadata\Resource`,

Reply:
looping through array is way better idea.